step/step.go

```diff
@@ -0,0 +1,200 @@
+	r.logger.Infof("Collecting xcresult files")
+
+	paths, err := collectXcresultFiles(config.TestDeployDir)
+	if err != nil {
+		return Result{}, fmt.Errorf("failed to find all xcresult files: %w", err)
+	}
+
+	r.logger.Printf("List of files:")
+	for _, path := range paths {
+		r.logger.Printf("- %s", path)
+	}
+
+	rootDir, err := testReportsRootDir()
+	if err != nil {
+		return Result{}, fmt.Errorf("failed to create test report directory: %w", err)
+	}
+
+	r.logger.Println()
+	r.logger.Infof("Generating reports")
+
+	for _, path := range paths {
+		if err := r.generateTestReport(rootDir, path); err != nil {
+			r.logger.Errorf("failed to generate test report (%s): %w", path, err)
+		}
+	}
+
+	r.logger.Println()
+	r.logger.Donef("Finished")
+
+	return Result{
+		TestReportDir: rootDir,
+	}, nil
+}
+
+func (r *ReportGenerator) Export(result Result) error {
+	return r.exporter.ExportOutput(testReportDirKey, result.TestReportDir)
+}
+
+func (r *ReportGenerator) generateTestReport(rootDir string, xcresultPath string) error {
+	baseName := strings.TrimSuffix(filepath.Base(xcresultPath), filepath.Ext(xcresultPath))
+	dirPath := filepath.Join(rootDir, baseName)
+	err := os.Mkdir(dirPath, 0755)
+	if err != nil {
+		if os.IsExist(err) {
+			r.logger.Warnf("Html report already exists for %s at %s", baseName, dirPath)
+			return nil
+		}
+		return err
+	}
+
+	params := []string{"run", "bitrise-io/XCTestHTMLReport@speed-improvements", "--output", dirPath, xcresultPath}
+	cmd := r.commandFactory.Create("mint", params, nil)
+	_, err = cmd.RunAndReturnTrimmedCombinedOutput()
+	if err != nil {
+		return fmt.Errorf("failed to export html: %w", err)
+	}
+
+	if err := moveAssets(xcresultPath, dirPath); err != nil {
+		return fmt.Errorf("failed to move assets: %w", err)
+	}
+
+	return nil
+}
+
+func collectXcresultFiles(dir string) ([]string, error) {
+	var files []string
+
+	err := filepath.WalkDir(dir, func(path string, d fs.DirEntry, err error) error {
+		if err != nil {
+			return err
+		}
+
+		isDir := d.IsDir()
+		extension := filepath.Ext(d.Name())
+
+		if isDir && extension == ".xcresult" {
+			files = append(files, path)
+		}
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return files, nil
+}
+
+func testReportsRootDir() (string, error) {
+	return os.MkdirTemp("", "test-reports")
+}
+
+func moveAssets(xcresultPath string, testReportDir string) error {
+	entries, err := os.ReadDir(xcresultPath)
+	if err != nil {
+		return err
+	}
+
+	assetFolder := filepath.Join(testReportDir, filepath.Base(xcresultPath))
+	if err := os.Mkdir(assetFolder, 0755); err != nil {
+		return err
+	}
+
+	for _, entry := range entries {
+		// The assets are dumped into the root, so we do not need the folders.
+		if entry.IsDir() {
+			continue
+		}
+
+		extension := filepath.Ext(entry.Name())
+		// We want to only move the useful assets which are images and videos.
+		if extension == ".plist" || extension == ".log" {
```

Comment:
wouldn't it be more safe if we moved what we know needs to be moved?

Reply:
It might not. I checked Xcode 14 and 15 xcresult files and the plist file is part of the xcresult bundle and the log file is generated by the tool we use. So we know what these two are.

But user can attach png, jpg, mp4 and arbitrary data files as attachments. If we by mistake do not upload one of them then they will not be able to access it from the html. And we do not know what the users will attach to their tests. So my idea was to only filter out which we know for sure that should not be there. Everything else will be uploaded. Worst case we upload something that should not be. But in that case mostly probably the html file will not have a link to it so the user will have no way to access it. And these artefacts will have a short expiry date after which they will be cleaned up.